x/upgrade/keeper.go

```diff
@@ -0,0 +1,181 @@
+)
+
+// NewKeeper constructs an upgrade keeper
+func NewKeeper(storeKey sdk.StoreKey, cdc *codec.Codec) Keeper {
+	return &keeper{
+		storeKey:        storeKey,
+		cdc:             cdc,
+		upgradeHandlers: map[string]Handler{},
+	}
+}
+
+func (keeper *keeper) SetUpgradeHandler(name string, upgradeHandler Handler) {
+	keeper.upgradeHandlers[name] = upgradeHandler
+}
+
+func (keeper *keeper) SetWillUpgrader(willUpgrader func(ctx sdk.Context, plan Plan)) {
+	keeper.willUpgrader = willUpgrader
+}
+
+func (keeper *keeper) SetOnUpgrader(onUpgrader func(ctx sdk.Context, plan Plan)) {
+	keeper.onUpgrader = onUpgrader
+}
+
+func (keeper *keeper) ScheduleUpgrade(ctx sdk.Context, plan Plan) sdk.Error {
+	err := plan.ValidateBasic()
+	if err != nil {
+		return err
+	}
+	if !plan.Time.IsZero() {
+		if !plan.Time.After(ctx.BlockHeader().Time) {
+			return sdk.ErrUnknownRequest("Upgrade cannot be scheduled in the past")
+		}
+		if plan.Height != 0 {
+			return sdk.ErrUnknownRequest("Only one of Time or Height should be specified")
+		}
+	} else {
+		if plan.Height <= ctx.BlockHeight() {
+			return sdk.ErrUnknownRequest("Upgrade cannot be scheduled in the past")
+		}
+	}
+	store := ctx.KVStore(keeper.storeKey)
+	if store.Has(DoneHeightKey(plan.Name)) {
+		return sdk.ErrUnknownRequest(fmt.Sprintf("Upgrade with name %s has already been completed", plan.Name))
+	}
+	bz := keeper.cdc.MustMarshalBinaryBare(plan)
+	keeper.haveCache = false
+	store.Set([]byte(PlanKey), bz)
+	return nil
+}
+
+func (keeper *keeper) ClearUpgradePlan(ctx sdk.Context) {
+	store := ctx.KVStore(keeper.storeKey)
+	keeper.haveCache = false
+	store.Delete([]byte(PlanKey))
+}
+
+// ValidateBasic does basic validation of a Plan
+func (plan Plan) ValidateBasic() sdk.Error {
+	if len(plan.Name) == 0 {
+		return sdk.ErrUnknownRequest("Name cannot be empty")
+
+	}
+	return nil
+}
+
+func (keeper *keeper) GetUpgradePlan(ctx sdk.Context) (plan Plan, havePlan bool) {
+	store := ctx.KVStore(keeper.storeKey)
+	bz := store.Get([]byte(PlanKey))
+	if bz == nil {
+		return plan, false
+	}
+	keeper.cdc.MustUnmarshalBinaryBare(bz, &plan)
+	return plan, true
+}
+
+// DoneHeightKey returns a key that points to the height at which a past upgrade was applied
+func DoneHeightKey(name string) []byte {
+	return []byte(fmt.Sprintf("done/%s", name))
+}
+
+func (keeper *keeper) BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock) {
+	blockTime := ctx.BlockHeader().Time
+	blockHeight := ctx.BlockHeight()
+
+	if !keeper.haveCache {
+		plan, found := keeper.GetUpgradePlan(ctx)
+		keeper.haveCachedPlan = found
+		keeper.plan = plan
+		keeper.haveCache = true
+		if found {
+			willUpgrader := keeper.willUpgrader
+			if willUpgrader != nil {
+				willUpgrader(ctx, keeper.plan)
+			}
+		}
+	}
+
+	if !keeper.haveCachedPlan {
+		return
+	}
+
+	upgradeTime := keeper.plan.Time
+	upgradeHeight := keeper.plan.Height
+	if (!upgradeTime.IsZero() && !blockTime.Before(upgradeTime)) || upgradeHeight <= blockHeight {
```

Comment:
Nice logic. I think the feature switch needs to be more integrated in the rest of the framework.